src/client/Microsoft.Identity.Client/Instance/Discovery/NetworkMetadataProvider.cs

```diff
@@ -78,20 +81,38 @@ private void CacheInstanceDiscoveryMetadata(InstanceDiscoveryResponse instanceDi
             client.AddQueryParameter("api-version", "1.1");
             client.AddQueryParameter("authorization_endpoint", BuildAuthorizeEndpoint(authority));
 
+            Uri instanceDiscoveryEndpoint = ComputeHttpEndpoint(authority, requestContext);
+
+            InstanceDiscoveryResponse discoveryResponse = await client
+                .DiscoverAadInstanceAsync(instanceDiscoveryEndpoint, requestContext)
+                .ConfigureAwait(false);
+
+            return discoveryResponse;
+        }
+
+        private Uri ComputeHttpEndpoint(Uri authority, RequestContext requestContext)
+        {
+            if (_userProvidedInstanceDiscoveryUri != null)
+            {
+                return _userProvidedInstanceDiscoveryUri;
```

Comment:
just a question, but should we be checking at all that the uri provided actually works, or we assume the developer has done due diligence?

Reply:
Wondering what checks we can do? We do an actual call to this end-point right after computing it and that's where we will find out whether it works or not, I believe. Did you have other things in mind?

---
In reply to: [390664823](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1689#discussion_r390664823) [](ancestors = 390664823)

Reply:
In this implementation, I simply replace the computed instance uri with a user given uri. All the optimizations that we've made in the past around discovery still apply, but when MSAL needs to access the endpoint, it goes to this custom uri instead of calculating its own uri. 

JFYI, our current logic states that unless validate authority is set to true, then MSAL will ignore any discovery errors (e.g. instance endpoint is down). This was done to increase availability.